Userland/Applications/SoundPlayer/PlaylistWidget.cpp

```diff
@@ -0,0 +1,127 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "PlaylistWidget.h"
+#include "Player.h"
+#include "SoundPlayerWidgetAdvancedView.h"
+#include <AK/LexicalPath.h>
+#include <LibGUI/BoxLayout.h>
+#include <LibGUI/HeaderView.h>
+#include <LibGUI/Model.h>
+#include <LibGUI/Window.h>
+
+PlaylistWidget::PlaylistWidget()
+{
+    set_layout<GUI::VerticalBoxLayout>();
+    m_table_view = add<PlaylistTableView>();
+    m_table_view->set_selection_mode(GUI::AbstractView::SelectionMode::SingleSelection);
+    m_table_view->set_selection_behavior(GUI::AbstractView::SelectionBehavior::SelectRows);
+    m_table_view->set_highlight_selected_rows(true);
+    m_table_view->on_doubleclick = [&](const Gfx::Point<int>& point) {
+        auto player = dynamic_cast<Player*>(window()->main_widget());
+        auto index = m_table_view->index_at_event_position(point);
+        if (!index.is_valid())
+            return;
+        player->open_file(m_table_view->model()->data(index, static_cast<GUI::ModelRole>(0x101)).as_string());
+    };
+}
+
+GUI::Variant PlaylistModel::data(const GUI::ModelIndex& index, GUI::ModelRole role) const
+{
+    if (role == GUI::ModelRole::TextAlignment)
+        return "CenterLeft";
+    if (role == GUI::ModelRole::Display) {
+        switch (index.column()) {
+        case 0:
+            return m_playlist_items[index.row()].extended_info->track_display_title.value_or(LexicalPath(m_playlist_items[index.row()].path).title());
+        case 1:
+            return format_duration(m_playlist_items[index.row()].extended_info->track_length_in_seconds.value_or(0));
+        case 2:
+            return m_playlist_items[index.row()].extended_info->group_name.value_or("");
+        case 3:
+            return m_playlist_items[index.row()].extended_info->album_title.value_or("");
+        case 4:
+            return m_playlist_items[index.row()].extended_info->album_artist.value_or("");
+        case 5:
+            return format_filesize(m_playlist_items[index.row()].extended_info->file_size_in_bytes.value_or(0));
+        }
+    }
+    if (role == GUI::ModelRole::Sort)
+        return data(index, GUI::ModelRole::Display);
+    if ((int)role == 0x101) //path
+        return m_playlist_items[index.row()].path;
```

Comment:
That feels a bit weeird, hardcoded like that. [`AutocompleteProvider.cpp`](https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibGUI/AutocompleteProvider.cpp#L84) seems to be the only other file that uses extended roles, and it still adds [a new `InternalRoles` enum](https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibGUI/AutocompleteProvider.cpp#L52). How about doing something like this ?

Reply:
Yeah, fixed :)